### PR TITLE
create a new errChan struct that wraps an err chan

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- The err chan that was used during the snapshotting process was prone to incorrect usage.
  
## Proposed changes
- Create a new struct that wraps the err chan. This new struct will prevent the wrongful usage of the err chan. 

## Testing procedure
- Normal testing procedure.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: state/accountsDB.go [hunk 1]
```diff
@@ -1283,15 +1284,15 @@ func (adb *AccountsDB) processSnapshotCompletion(
...
+func (adb *AccountsDB) syncMissingNodes(missingNodesChan chan []byte, errChan common.BufferedErrChan, stats *snapshotStatistics, syncer AccountsDBSyncer) {
 	defer stats.SyncFinished()
 
 	if check.IfNil(syncer) {
 		log.Error("can not sync missing nodes", "error", ErrNilTrieSyncer.Error())
 		for missingNode := range missingNodesChan {
 			log.Warn("could not sync node", "hash", missingNode)
 		}
-		errChan <- ErrNilTrieSyncer
+		errChan.WriteInChanNonBlocking(ErrNilTrieSyncer)
```
Comment (iulianpascalau) [inline state/accountsDB.go:1295]: all these blocking calls became non-blocking calls. I guess this is desirable.

File: trie/patriciaMerkleTrie.go [hunk 2]
```diff
@@ -481,7 +481,7 @@ func (tr *patriciaMerkleTrie) GetAllLeavesOnChannel(
 			ctx,
 		)
 		if err != nil {
-			writeInChanNonBlocking(leavesChannels.ErrChan, err)
+			leavesChannels.ErrChan.WriteInChanNonBlocking(err)
```
Comment (iulianpascalau) [inline trie/patriciaMerkleTrie.go:484]: the old `writeInChanNonBlocking` method is now used only in tests. Move the implementation directly in the export_test.go file (in WriteInChanNonBlocking function)?

File: trie/trieStorageManager_test.go [hunk 3]
```diff
@@ -36,6 +37,12 @@ func getNewTrieStorageManagerArgs() trie.NewTrieStorageManagerArgs {
 	}
 }
 
+// ErrChanWithLen extends the BufferedErrChan interface with a Len method
+type ErrChanWithLen interface {
```
Comment (iulianpascalau) [inline trie/trieStorageManager_test.go:41]: If we need this only in tests, we might change this interface to unexported.

File: common/errChan/errChan.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package errChan
+
+import "sync"
+
+type errChan struct {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/errChan/errChan.go:5]: Done

File: common/errChan/errChan.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	return &errChan{
+		ch:     make(chan error, 1),
+		closed: false,
+	}
+}
+
+// WriteInChanNonBlocking will send the given error on the channel if the chan is not blocked
+func (ec *errChan) WriteInChanNonBlocking(err error) {
+	select {
+	case ec.ch <- err:
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/errChan/errChan.go:22]: Right, fixed and added tests.

File: common/errChan/errChan_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+	ec := NewErrChan()
+	assert.Equal(t, 0, ec.Len())
+
+	ec.ch <- fmt.Errorf("err1")
+	assert.Equal(t, 1, ec.Len())
+
+	ec.WriteInChanNonBlocking(fmt.Errorf("err2"))
+	assert.Equal(t, 1, ec.Len())
+}
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/errChan/errChan_test.go:81]: Added.

File: state/accountsDB.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply state/accountsDB.go:1295]: Yes, this was the bug that prompted this refactoring.

File: trie/patriciaMerkleTrie.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:484]: Done.

File: trie/trieStorageManager_test.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager_test.go:41]: Done.

**Review by iulianpascalau — DISMISSED** [review]

File: epochStart/metachain/systemSCs.go [hunk 7]
```diff
@@ -1125,7 +1126,7 @@ func (s *systemSCProcessor) getArgumentsForSetOwnerFunctionality(userValidatorAc
 		}
 	}
 
-	err = common.GetErrorFromChanNonBlocking(leavesChannels.ErrChan)
```
Comment (ssd04) [inline epochStart/metachain/systemSCs.go:1128]: i think `GetErrorFromChanNonBlocking` from common is not used anymore, maybe we can remove it

**Review by ssd04 — APPROVED** [review]

File: epochStart/metachain/systemSCs.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply epochStart/metachain/systemSCs.go:1128]: Removed.

**Review by ssd04 — APPROVED** [review]

**Review by gabi-vuls — DISMISSED** [review]
System test passed .

@@                            Log scanner                                  @@
##                            refactor-err-chan                            ##
================================================================================
- Known Warnings              12
- New Warnings                3
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================
- block hash does not match                  15789
- wrong nonce in block                       4257
- miniblocks does not match                  0
- num miniblocks does not match              0
- miniblock hash does not match              0
- block bodies does not match                0
- receipts hash missmatch                    1
================================================================================
-  No jailed nodes on the thestnet
================================================================================

**Review by iulianpascalau — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**